src/main/scala/state/PreconditionPropagationTransformer.scala

```diff
@@ -0,0 +1,36 @@
+package viper.silicon.state
+
+import viper.silicon.rules.functionSupporter
+import viper.silicon.state.terms.{And, App, HeapDepFun, Implies, Ite, Let, Literal, Not, Or, Quantification, Term, True}
+import viper.silver.ast
+
+object PreconditionPropagationTransformer {
+  def transform(t: Term, p: ast.Program): Term = {
+    val res = t match {
+      case _:Literal => True()
+      case And(ts) => And(transform(ts.head, p), Implies(ts.head, transform(And(ts.tail), p)))
```

Comment:
AFAICS, this transforms A && B into A' && (A ==> B'), but it's not clear to me, why (and similar for later cases). Would be nice (although somewhat non-traditional) to have a brief description of what the transformer does.

Reply:
Yes. It's because when A&&B is checked for well-definedness, because of short-circuiting, we check that A is well-defined, and if A is true, then we also check if B is well- defined. Here, we're assuming exactly that. Similar argument for other cases. 
I'll add some documentation explaining this.